### PR TITLE
Alerting: Enable sending notifications to a specific topic on Telegram

### PR DESCRIPTION
**Pull request by th0th**
I need help, please check the **Special notes for your reviewer** section.

**What is this feature?**

Telegram has a feature called [group topics](https://telegram.org/blog/topics-in-groups-collectible-usernames/it?setln=en#topics-in-groups), which is like a channel in Slack or Discord. Current Telegram alerting mechanism doesn't allow sending notifications to a specific topic. This PR aims to add (optionally) setting the topic for Telegram type contact point.

**Why do we need this feature?**

It is very common for Telegram bot integrations to set a custom topic for the messages sent. For example, a group can have multiple topics like #general #notifications, and the user might want to send alerts to the #notifications topic.

**Who is this feature for?**

For the folks who use Grafana with Telegram alerts.

**Special notes for your reviewer:**

I added the required changes for the feature, however, I couldn't get it working, so I need a fresh pair of eyes :) It is like some of the changes I made on my development environment are not taken into consideration. For example, I tried replacing all the `api.telegram.org`s in the code with a random, invalid URL, but the "Test" button on the "Contact points" still successfully sent the message to Telegram.

Once we get the thing working, I am planning to update relevant part of the docs as well 👍 

Please check that:
- [ ] It works as expected from a user's perspective.
- [x] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

**Comment by yuri-tceretian** [comment]
Hi @thoth! I appreciate your contribution! However, currently, we do not accept any changes to the integrations that do not exist in the Prometheus Alertmanager (see https://prometheus.io/docs/alerting/latest/configuration/#telegram_config) because we are trying to unify the Grafana and Prometheus Alertmanagers.
I would recommend the following:
1. Open a PR in https://github.com/prometheus/alertmanager to add this field to [TelegramConfig](https://github.com/prometheus/alertmanager/blob/main/config/notifiers.go#L747-L760)
2. Open a PR in https://github.com/grafana/alerting/issues to add this field to [telegram.Config](https://github.com/grafana/alerting/blob/main/receivers/telegram/config.go). This is needed anyway to make this PR working because here it's only the API models.
3. Ping me about the PRs, and I will make sure they are reviewed promptly. 

4. Once those 2 PRs are merged, we will be able to merge this one.
5. After all 4 PRs are merged, we will need to update Terraform provider . Open a PR to add this field to the Telegram [resource ](https://github.com/grafana/terraform-provider-grafana/blob/master/internal/resources/grafana/resource_alerting_contact_point_notifiers.go#L1629-L1669) in the Grafana Terraform Provider.

**Comment by thoth** [comment]
I think you sent this to the wrong account...

**Comment by th0th** [comment]
> Hi @thoth! I appreciate your contribution! However, currently, we do not accept any changes to the integrations that do not exist in the Prometheus Alertmanager (see https://prometheus.io/docs/alerting/latest/configuration/#telegram_config) because we are trying to unify the Grafana and Prometheus Alertmanagers. I would recommend the following:
> 
> 1. Open a PR in https://github.com/prometheus/alertmanager to add this field to [TelegramConfig](https://github.com/prometheus/alertmanager/blob/main/config/notifiers.go#L747-L760)
> 2. Open a PR in https://github.com/grafana/alerting/issues to add this field to [telegram.Config](https://github.com/grafana/alerting/blob/main/receivers/telegram/config.go). This is needed anyway to make this PR working because here it's only the API models.
> 3. Ping me about the PRs, and I will make sure they are reviewed promptly.
> 4. Once those 2 PRs are merged, we will be able to merge this one.
> 5. After all 4 PRs are merged, we will need to update Terraform provider . Open a PR to add this field to the Telegram [resource ](https://github.com/grafana/terraform-provider-grafana/blob/master/internal/resources/grafana/resource_alerting_contact_point_notifiers.go#L1629-L1669) in the Grafana Terraform Provider.
> 
> Also, there is another PR that adds a similar field #79534. I left the same message there.

Hey @yuri-tceretian, thanks for the explanation.

I created these 3 PRs:

* Prometheus Alertmanager: https://github.com/prometheus/alertmanager/pull/3638
* alerting: https://github.com/prometheus/alertmanager/pull/3638
* Terraform provider: https://github.com/grafana/terraform-provider-grafana/pull/1234

I didn't quite understand how all these are connected to each other, is there is a way I can run this implementation on my local machine, please tell me and I will test :) Also, let me know if there are additional changes needed 💐

**Comment by yuri-tceretian** [comment]
> I think you sent this to the wrong account...

Oh, sorry :)

**Comment by th0th** [comment]
> ```
> replace github.com/grafana/alerting => ../alerting
> ```

Thank you, do I need to do the same thing about `alertmanager`, too? There is already a `replace` line for it: https://github.com/grafana/grafana/blob/main/go.mod#L498

**Comment by yuri-tceretian** [comment]
you can replace that "replace" with your version.

**Comment by th0th** [comment]
> you can replace that "replace" with your version.

I replaced the `alertmanager` with my local clone, but it didn't work. Looks like there is a difference between grafana's fork and the upstream one. Should we implement this Telegram change thing on https://github.com/grafana/prometheus-alertmanager as well?

```
go build -ldflags -w -X main.version=10.3.0-pre -X main.commit=88d47e3257 -X main.buildstamp=1702592018 -X main.buildBranch=telegram-message-thread-id -o ./bin/grafana ./pkg/cmd/grafana
# github.com/grafana/grafana/pkg/services/ngalert/api/tooling/definitions
pkg/services/ngalert/api/tooling/definitions/alertmanager.go:480:25: undefined: amv2.Integration
# github.com/grafana/alerting/notify
../alerting/notify/grafana_alertmanager.go:48:10: undefined: silence.ValidateMatcher
../alerting/notify/grafana_alertmanager.go:108:28: undefined: notify.Receiver
../alerting/notify/grafana_alertmanager.go:149:30: undefined: notify.Receiver
../alerting/notify/grafana_alertmanager.go:412:26: undefined: notify.Receiver
../alerting/notify/grafana_alertmanager.go:419:40: undefined: notify.NewReceiver
../alerting/notify/grafana_alertmanager.go:445:71: undefined: notify.Receiver
../alerting/notify/grafana_alertmanager.go:621:38: cannot use integrations[i] (variable of type *"github.com/prometheus/alertmanager/notify".Integration) as "github.com/prometheus/alertmanager/notify".Integration value in argument to notify.NewRetryStage
../alerting/notify/factory.go:59:40: cannot use i (variable of type "github.com/prometheus/alertmanager/notify".Integration) as *"github.com/prometheus/alertmanager/notify".Integration value in argument to append
../alerting/notify/grafana_alertmanager_metrics.go:25:44: too many arguments in call to metrics.NewAlerts
	have (string, prometheus.Registerer)
	want (prometheus.Registerer)
exit status 1
exit status 1
make[1]: *** [build-go] Error 1
```

**Comment by yuri-tceretian** [comment]
To test your two PRs in Grafana and Alerting repositories, you can checkout corresponding branches in both repositories.
Then open `go.mod` and add the following at the bottom of the file
```
replace github.com/grafana/alerting => <path_to_alerting_repo_dir>
```
In my case, when two repositories are in the same folder the line is 
```
replace github.com/grafana/alerting => ../alerting
```

Now, you can build Grafana with your version of alerting repository.

**Comment by yuri-tceretian** [comment]
>I replaced the alertmanager with my local clone, but it didn't work. Looks like there is a difference between grafana's fork and the upstream one. Should we implement this Telegram change thing on https://github.com/grafana/prometheus-alertmanager as well?

Right, we use forked version of Prometheus Alertmanager and synchronize it with upstream regularly. So, you do not need to manually port it to the fork. Also, you do not need to import the Alertmanager PR to Grafana. There is a weak relationship between Prometheus Alertmanager and Grafana. Grafana provides a UI for it, and it is configured as an external service, via Grafana data source.
Moreover, the Prometheus Alertmanager does not support configuration via API. The only way to configure it is via YAML file. Therefore, Grafana only shows the config in read-only mode. However, Alertmanager that is a part of Grafnaa Mimir does let you configure via API and Grafana supports. 
That being said, I would recommend you treat the PR in Alertmanager as a separate part from this PR and the one in alerting repository. It is required to keep the diff between flavors of Alertmanager minimal but it is not required to support it in Grafana.

File: pkg/services/ngalert/notifier/channels_config/available_channels.go
Comment (yuri-tceretian) [inline pkg/services/ngalert/notifier/channels_config/available_channels.go:825]: let's add `ValidationRule`, a regular expression that makes sure that the input is an integer. Something like `^-?[0-9]{19}$`

**Review by gillesdemey — COMMENTED** [review]
LGTM from the front-end side, but will defer approval to the BE

File: pkg/services/ngalert/notifier/channels_config/available_channels.go
Comment (th0th) [reply pkg/services/ngalert/notifier/channels_config/available_channels.go:826]: Done 👍

File: public/app/features/alerting/unified/utils/cloud-alertmanager-notifier-types.ts
Comment (yuri-tceretian) [inline public/app/features/alerting/unified/utils/cloud-alertmanager-notifier-types.ts:390]: this can be removed until PR in Alertmanager is merged.

File: pkg/services/alerting/notifiers/telegram.go
Comment (yuri-tceretian) [inline pkg/services/alerting/notifiers/telegram.go:58]: this is not needed as it is related to the legacy alerting that will be deleted soon

File: pkg/services/ngalert/api/tooling/definitions/contact_points.go
Comment (yuri-tceretian) [inline pkg/services/ngalert/api/tooling/definitions/contact_points.go:193]: please use correct tags for JSON and YAML, i.e `message_thread_id`

File: pkg/services/ngalert/notifier/channels_config/available_channels.go
Comment (yuri-tceretian) [inline pkg/services/ngalert/notifier/channels_config/available_channels.go:823]: This field name should match https://github.com/grafana/alerting/pull/143/files#diff-ed9e160428d5b291f759bad53d5b04e320378f97a0340630c18041357fa44fa2R23

**Review by yuri-tceretian — DISMISSED** [review]
I approved one in alerting repo but this one needs some work.

File: pkg/services/ngalert/notifier/channels_config/available_channels.go
Comment (th0th) [reply pkg/services/ngalert/notifier/channels_config/available_channels.go]: Done.

File: public/app/features/alerting/unified/utils/cloud-alertmanager-notifier-types.ts
Comment (th0th) [reply public/app/features/alerting/unified/utils/cloud-alertmanager-notifier-types.ts]: I removed this part, but didn't fully understand what you meant. Should I open another PR later for this part?

File: pkg/services/alerting/notifiers/telegram.go
Comment (th0th) [reply pkg/services/alerting/notifiers/telegram.go]: I reverted the file 👍

File: pkg/services/ngalert/api/tooling/definitions/contact_points.go
Comment (th0th) [reply pkg/services/ngalert/api/tooling/definitions/contact_points.go]: Done.

**Comment by th0th** [comment]
@yuri-tceretian I fixed stuff you mentioned. Can you please have another look?

**Review by yuri-tceretian — COMMENTED** [review]
I merged PR in alerting repository. So, now you can execute
```
go get github.com/grafana/alerting@ec2c02267fa591e7252ddb44143555bc941ce6fa
```
and commit those changes into your PR.  Unfortunately, your PR seems to be blocked by PR https://github.com/grafana/grafana/pull/81512 because alerting module contains some changes that have not updated in Grafana. Let's wait for that PR merged first.

**Review by yuri-tceretian — APPROVED** [review]
LGTM

**Comment by th0th** [comment]
Thank you @yuri-tceretian 💐

**Pull request merged**
